src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,770 @@
+        HandleRouterAdvertisement(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    case Ip6::Icmp::Header::kTypeRouterSolicit:
+        HandleRouterSolicit(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    default:
+        ExitNow();
+    }
+
+exit:
+    return;
+}
+
+void RoutingManager::HandleNotifierEvents(Events aEvents)
+{
+    VerifyOrExit(IsInitialized());
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
+    }
+
+    if (aEvents.Contains(kEventThreadNetdataChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            EvaluateRoutingPolicy();
+        }
+    }
+
+exit:
+    return;
+}
+
+uint8_t RoutingManager::EvaluateOmrPrefix(Ip6::Prefix *aNewOmrPrefixes, uint8_t aMaxOmrPrefixNum)
+{
+    uint8_t                         newOmrPrefixNum = 0;
+    NetworkData::Iterator           iterator        = NetworkData::kIteratorInit;
+    NetworkData::OnMeshPrefixConfig onMeshPrefixConfig;
+    Ip6::Prefix *                   smallestOmrPrefix       = nullptr;
+    Ip6::Prefix *                   publishedLocalOmrPrefix = nullptr;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, onMeshPrefixConfig) == OT_ERROR_NONE)
+    {
+        uint8_t newPrefixIndex;
+
+        if (!IsValidOmrPrefix(onMeshPrefixConfig.GetPrefix()) || !onMeshPrefixConfig.mDefaultRoute ||
+            !onMeshPrefixConfig.mSlaac || onMeshPrefixConfig.mDp)
+        {
+            continue;
+        }
+
+        newPrefixIndex = 0;
+        while (newPrefixIndex < newOmrPrefixNum && onMeshPrefixConfig.GetPrefix() != aNewOmrPrefixes[newPrefixIndex])
+        {
+            ++newPrefixIndex;
+        }
+
+        if (newPrefixIndex != newOmrPrefixNum)
+        {
+            // Ignore duplicate prefixes.
+            continue;
+        }
+
+        if (newOmrPrefixNum >= aMaxOmrPrefixNum)
+        {
+            otLogWarnBr("EvaluateOmrPrefix: too many OMR prefixes, ignoring prefix %s",
+                        onMeshPrefixConfig.GetPrefix().ToString().AsCString());
+            continue;
+        }
+
+        aNewOmrPrefixes[newOmrPrefixNum] = onMeshPrefixConfig.GetPrefix();
+        if (smallestOmrPrefix == nullptr || onMeshPrefixConfig.GetPrefix() < *smallestOmrPrefix)
+        {
+            smallestOmrPrefix = &aNewOmrPrefixes[newOmrPrefixNum];
+        }
+        if (aNewOmrPrefixes[newOmrPrefixNum] == mLocalOmrPrefix)
+        {
+            publishedLocalOmrPrefix = &aNewOmrPrefixes[newOmrPrefixNum];
+        }
+        ++newOmrPrefixNum;
+    }
+
+    // Decide if we need to add or remove my local OMR prefix.
+    if (newOmrPrefixNum == 0)
+    {
+        otLogInfoBr("EvaluateOmrPrefix: no valid OMR prefixes found in Thread network");
+        if (PublishLocalOmrPrefix() == OT_ERROR_NONE)
```

Comment:
It would be much better to add a random delay after Network data change and publishing local OMR prefix to reduce race conditions between multiple BRs.

From implementation's perspective, I would suggest to add a random delay before invoking `EvaluateRoutingPolicy` after `kEventThreadNetdataChanged` is triggered.

Reply:
SGTM. But prefer adding it in separate PR.